tests/test_mla_decode_kernel.py

```diff
@@ -0,0 +1,353 @@
+                raise ValueError("For simplicity, kv_len should be multiple of page_size.")
+            num_pages_per_seq = kv_len // page_size
+            total_num_pages = num_pages_per_seq * bsz
+            
+            kv_indptr = torch.arange(0, bsz + 1).to(dev_id).int() * num_pages_per_seq
+            kv_indices = torch.arange(0, total_num_pages).to(dev_id).int()
+            kv_last_page_len = torch.full(
+                (bsz,), page_size, dtype=torch.int32
+            ).to(dev_id)
+            
+            paged_ckv_cache = compressed_kv_normed_cache.to(q_kv_dtype).reshape(total_num_pages, page_size, self.kv_lora_rank)
+            paged_kpe_cache = k_pe_cache.to(q_kv_dtype).reshape(total_num_pages, page_size, self.qk_rope_head_dim)
+            
+            workspace_buffer = torch.empty(32 * 1024 * 1024, dtype=torch.int8).to(dev_id)
+            wrapper = flashinfer.BatchDecodeMlaWithPagedKVCacheWrapper(workspace_buffer)
+            wrapper.plan(
+                kv_indptr,
+                kv_indices,
+                kv_last_page_len,
+                num_qo_heads=self.num_heads,
+                head_dim_compressed_kv=self.kv_lora_rank,
+                page_size=page_size,
+                sm_scale=self.softmax_scale,
+                rope_theta=self.rope_theta,
+                data_type=q_kv_dtype,
+                q_data_type=q_kv_dtype)
+            attn_output = wrapper.run(q_nope, q_pe, paged_ckv_cache, paged_kpe_cache)
+            
+            attn_output = attn_output.to(self.W_UV_O.dtype)
+        
+        # output ~ [bsz, 5120]
+        output = torch.matmul(
+                attn_output.reshape(bsz, self.num_heads*self.kv_lora_rank), 
+                self.W_UV_O) # W_UV_O ~ [65536, 5120]
+
+        return output
+
+
+if __name__ == "__main__":
+    
+    dev_id = 1
+    
+    torch.manual_seed(666)
+    torch.set_grad_enabled(False)
+    
+    mla_vanilla = DeepseekV2AttentionVanilla().cuda(device=dev_id)
+
+    bsz = 6
+    kv_len = 640
+    page_size = 16
+    
+    hidden_states = torch.randn([bsz, 1, mla_vanilla.hidden_size]).to(dev_id)
+    compressed_kv_normed_cache = torch.randn([bsz, kv_len, mla_vanilla.kv_lora_rank]).to(dev_id)
+    k_pe_cache = torch.randn([bsz, kv_len, mla_vanilla.qk_rope_head_dim]).to(dev_id)
+    
+    output_vanilla = mla_vanilla.run_decode(hidden_states, compressed_kv_normed_cache, k_pe_cache)
+    
+    mla_mat_absorb = DeepseekV2AttentionMatAbsorbDecode(mla_vanilla).cuda(device=dev_id)
+    output_mat_absorbed_use_torch = mla_mat_absorb.run_proof_of_concept(hidden_states.squeeze(1), compressed_kv_normed_cache, k_pe_cache, use_flashinfer_kernel=False)
+    output_mat_absorbed_use_flashinfer = mla_mat_absorb.run_proof_of_concept(hidden_states.squeeze(1), compressed_kv_normed_cache, k_pe_cache, use_flashinfer_kernel=True)
+    
+    cos_sim_use_torch = F.cosine_similarity(output_vanilla.reshape(-1), output_mat_absorbed_use_torch.reshape(-1), dim=0)
```

Comment:
Why use cosine similar here?

Reply:
Actually during my development of decode kernel, I tried hard to align the kernel implementation with the correct computation semantic, I found this metric is a good indicator to guide me if I was in the right direction to fix the discrepancy, the value changed from 0.1 -> 0.95 -> 0.9999.
But now I just tried the MSE as the metric, it value is rather large than I expected, it seems that the cosine similarity is a relaxed standard.
So there are still some discrepancy, I will look into this issue.

Reply:
hi @yzh119 I just updated the test case, added both MSE and WMAPE metrics. Below is the output from one run, it seems that the f32 to f16 conversion can cause some precision loss. I still can't find wrong implementation in decode kernel.
```
cos_use_torch_f32 = 1.0
wmape_use_torch_f32 = 9.26729843382549e-06
mse_use_torch_f32=0.0008153514936566353

cos_use_torch_f16 = 0.9997764825820923
wmape_use_torch_f16 = 0.016850485358918303
mse_use_torch_f16 = 3793.904296875

cos_use_flashinfer = 0.9999209642410278
wmape_use_flashinfer = 0.012483024544464835
mse_use_flashinfer = 1346.939453125
```

The MSE value is rather large, because the elements in the output tensor are at the level of thousands, you can try it by yourself. 
Do you think WMAPE (corresponds to comparison of tensor's magnitude) and Cosine Similarity (corresponds to comparison of tensor's angle) together can be enough proof for algorithm correctness?

Reply:
Thanks for confirmation, I think cosine similarity is okay in this case.
Let's merge this PR first and investigate the numerical issue later.